src/main/java/net/iponweb/disthene/service/aggregate/AggregateService.java

```diff
@@ -0,0 +1,179 @@
+import net.iponweb.disthene.config.Rollup;
+import net.iponweb.disthene.events.DistheneEvent;
+import net.iponweb.disthene.events.MetricAggregateEvent;
+import net.iponweb.disthene.events.MetricStoreEvent;
+import net.iponweb.disthene.util.NamedThreadFactory;
+import org.apache.log4j.Logger;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author 
+ */
+@Listener(references= References.Strong)
+public class AggregateService {
+    private static final String SCHEDULER_NAME = "distheneAggregatorFlusher";
+    private static final int RATE = 10;
+    private volatile boolean shuttingDown = false;
+
+    private Logger logger = Logger.getLogger(AggregateService.class);
+
+    private MBassador<DistheneEvent> bus;
+    private DistheneConfiguration distheneConfiguration;
+    private Rollup baseRollup;
+
+    private ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1, new NamedThreadFactory(SCHEDULER_NAME));
+
+    private final ConcurrentNavigableMap<Long, ConcurrentMap<MetricKey, AverageRecord>> accumulator = new ConcurrentSkipListMap<>();
+
+    public AggregateService(MBassador<DistheneEvent> bus, DistheneConfiguration distheneConfiguration) {
+        this.distheneConfiguration = distheneConfiguration;
+        this.baseRollup = distheneConfiguration.getCarbon().getBaseRollup();
+        this.bus = bus;
+        bus.subscribe(this);
+
+        scheduler.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                flush();
+            }
+        }, 60 - ((System.currentTimeMillis() / 1000L) % 60), RATE, TimeUnit.SECONDS);
+    }
+
+    @Handler(rejectSubtypes = false)
+    public void handle(MetricAggregateEvent metricAggregateEvent) {
+        aggregate(metricAggregateEvent.getMetric());
+    }
+
+    private ConcurrentMap<MetricKey, AverageRecord> getTimestampMap(long timestamp) {
+        ConcurrentMap<MetricKey, AverageRecord> timestampMap = accumulator.get(timestamp);
+        if (timestampMap == null) {
+            ConcurrentMap<MetricKey, AverageRecord> newTimestampMap = new ConcurrentHashMap<>();
+            timestampMap = accumulator.putIfAbsent(timestamp, newTimestampMap);
+            if (timestampMap == null) {
+                timestampMap = newTimestampMap;
+            }
+        }
+
+        return timestampMap;
+    }
+
+    private AverageRecord getAverageRecord(ConcurrentMap<MetricKey, AverageRecord> map, MetricKey metricKey) {
+        AverageRecord averageRecord = map.get(metricKey);
+        if (averageRecord == null) {
+            AverageRecord newAverageRecord = new AverageRecord();
+            averageRecord = map.putIfAbsent(metricKey, newAverageRecord);
+            if (averageRecord == null) {
```

Comment:
another two ifs, hmm
How it's going here?
If there is no record, init a new one:
Why averageRecord is a map.put, and later newAverageRecord, why like that?

Reply:
fixed